### PR TITLE
Implement focus-follows-mouse

### LINKED ISSUE #9726: Prevent editor from stealing focus when link is clicked in terminal
### Check for existing issues

- [X] Completed

### Describe the feature

I tried to navigate to the terminal by right clicking on it to focus. I had several file paths displayed due to a previous drag. I accidentally clicked on one of the many links that caused the editor to steal focus, which I didn't notice and started typing my command.

I understand the convenience of having links open files directly in the editor, but perhaps the action should be "hidden" behind a `cmd-click` key combination instead, because in case most of the terminal is filled with links, it'd be tedious to find a place to focus the editor with the mouse.

### If applicable, add mockups / screenshots to help present your vision of the feature

https://github.com/zed-industries/zed/assets/16290753/dff770cb-bc32-4241-a8ef-f806049ec450

### PR DESCRIPTION
**Pull request by xyzqm**
Release Notes:

- Adds focus-follows-mouse to terminal

Optionally, include screenshots / media showcasing your addition that can be included in the release notes.

https://github.com/user-attachments/assets/7fd459e2-1b74-4d2c-bf46-7ba88c5e984c

- Fixes #8167 
- Fixes #9726

### CODE REVIEW

**Comment by mikayla-maki** [comment]
So, this PR removes the focus requirement on the mouse events in the terminal... I can get behind that change.

I'm not so sure about the global_key_event change, can you talk me through what that API is intended to facilitate, and how it works with this feature?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]
So, this PR removes the focus requirement on the mouse events in the terminal... I can get behind that change.

I'm not so sure about the global_key_event change, can you talk me through what that API is intended to facilitate, and how it works with this feature?

**Comment by danielzsh** [comment]
Originally, the terminal only received key events if it had focus, meaning if you did something like:
1. focus terminal and hold cmd
2. focus editor and release cmd
and went back to the terminal, it would still behave as though cmd were held down (meaning hovering any path will link to it, even without actually holding down cmd). Similarly, if you hold down the cmd key while in the editor and click into the terminal, the modifier will not register either. Perhaps this could be fixed by having the terminal access global modifier keys instead of maintaining its own modifier state, but I wasn't sure how to go about doing that.

So `global_key_change` just adds a key listener at the global level and no longer requires the element to be in focus for it to trigger.

**Comment by mikayla-maki** [comment]
Rather than patch over the bug with global_key_change, I think it would be better to fix the bug you describe here directly. Any `WindowContext` should have a `modifiers()` method you can use to grab the current global modifier state. Maybe from there, we can pass the modifiers down into `Terminal`, and remove the state?

**Comment by danielzsh** [comment]
I just realized WindowContext already has an `on_modifiers_changed` function, which should probably always register global listeners, so I'm considering just taking my logic from `global_key_change` and putting it into `on_modifiers_changed`; does that sound reasonable?

**Comment by mikayla-maki** [comment]
@danielzsh That does sound reasonable!

File: crates/gpui/src/key_dispatch.rs [hunk 1]
```diff
@@ -75,6 +75,7 @@ pub(crate) struct DispatchTree {
     view_node_ids: FxHashMap<EntityId, DispatchNodeId>,
     keymap: Rc<RefCell<Keymap>>,
     action_registry: Rc<ActionRegistry>,
+    modifiers_changed_listeners: Vec<DispatchNodeId>,
```
Comment (mikayla-maki) [inline crates/gpui/src/key_dispatch.rs:78]: I don't think this change is needed at all, modifiers changed events are already global.

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

File: crates/gpui/src/key_dispatch.rs (hunk 1 above)
Comment (danielzsh) [reply crates/gpui/src/key_dispatch.rs:78]: That's odd...when I get rid of `modifiers_changed_listeners`, the events are no longer global. I think a modifier event set directly on the window is global, but not one set through the DispatchTree?

Comment (mikayla-maki) [reply crates/gpui/src/key_dispatch.rs:78]: hmmm....

Comment (mikayla-maki) [reply crates/gpui/src/key_dispatch.rs:78]: I am definitely incorrect that modifiers changed events are global. In fact, none of the ones I thought where global currently are....

**Comment by mikayla-maki** [comment]
Oh I just checked the code and I've been totally wrong for a while here. There is no global modifiers changed actions!

**Comment by danielzsh** [comment]
So is this PR fine, or is there still something I should change (aside from resolving merge conflict)?

**Comment by danielzsh** [comment]
Hm, it looks like it's still not totally reponsive; sometimes the modifier state doesn't update properly, especially when you rapidly press and release the cmd key.

**Comment by mikayla-maki** [comment]
I'm sorry I've not followed up on this. That said, based on your last comment it looks like the current PR isn't in a good state, and hasn't been touched in a while. I'm going to close this for now, but feel free to open this again if you're still interested :)

**Pull request closed without merging**